sql/execstats: TestTraceAnalyzer data race

sql/execstats.TestTraceAnalyzer [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4212812&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4212812&tab=artifacts#/) on master @ [d5ed1d9cb363379731082ed9208b81feb39ab13a](https://github.com/cockroachdb/cockroach/commits/d5ed1d9cb363379731082ed9208b81feb39ab13a):


```
  github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigreconciler.(*incrementalReconciler).reconcile()
      /go/src/github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigreconciler/reconciler.go:359 +0x135
  github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigreconciler.(*Reconciler).Reconcile()
      /go/src/github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigreconciler/reconciler.go:173 +0x524
  github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigjob.(*resumer).Resume()
      /go/src/github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigjob/job.go:74 +0x487
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1123 +0x164
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1124 +0x8db
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:401 +0x74e
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:323 +0x125
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x235

Goroutine 34 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1504 +0x9d1
  github.com/cockroachdb/cockroach/pkg/sql/execstats_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execstats/main_test.go:28 +0x186
  main.main()
      _testmain.go:53 +0x264

Goroutine 367 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x615
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:354 +0xffb
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:320 +0xc7d
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeClaimedJobs.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:205 +0xd3
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeClaimedJobs·dwrap·1()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:208 +0x47
==================
=== CONT  TestTraceAnalyzer
    traceanalyzer_test.go:192: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTraceAnalyzer280679145
    testing.go:1152: race detected during execution of test
--- FAIL: TestTraceAnalyzer (4.42s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTraceAnalyzer.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/execstats.TestTraceAnalyzer [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4226198&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4226198&tab=artifacts#/) on master @ [c2023effaad7966ace04f4d84170f96dc47ce6ba](https://github.com/cockroachdb/cockroach/commits/c2023effaad7966ace04f4d84170f96dc47ce6ba):


```
  github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigreconciler.(*incrementalReconciler).reconcile()
      /go/src/github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigreconciler/reconciler.go:359 +0x135
  github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigreconciler.(*Reconciler).Reconcile()
      /go/src/github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigreconciler/reconciler.go:173 +0x524
  github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigjob.(*resumer).Resume()
      /go/src/github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigjob/job.go:74 +0x487
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1123 +0x164
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1124 +0x8db
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:401 +0x74e
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:323 +0x125
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x235

Goroutine 33 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1504 +0x9d1
  github.com/cockroachdb/cockroach/pkg/sql/execstats_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execstats/main_test.go:28 +0x186
  main.main()
      _testmain.go:53 +0x264

Goroutine 362 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x615
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:354 +0xffb
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:320 +0xc7d
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeClaimedJobs.func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:205 +0xd3
  github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeClaimedJobs·dwrap·1()
      /go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:208 +0x47
==================
=== CONT  TestTraceAnalyzer
    traceanalyzer_test.go:191: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTraceAnalyzer875976080
    testing.go:1152: race detected during execution of test
--- FAIL: TestTraceAnalyzer (4.54s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTraceAnalyzer.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/execstats/traceanalyzer_test.go
+++ b/pkg/sql/execstats/traceanalyzer_test.go
@@ -107,17 +107,14 @@ func TestTraceAnalyzer(t *testing.T) {
 	for _, vectorizeMode := range []sessiondatapb.VectorizeExecMode{sessiondatapb.VectorizeOff, sessiondatapb.VectorizeOn} {
 		execCtx, finishAndCollect := tracing.ContextWithRecordingSpan(ctx, execCfg.AmbientCtx.Tracer, t.Name())
 		defer finishAndCollect()
-		ie := execCfg.InternalExecutor
-		ie.SetSessionData(
-			&sessiondata.SessionData{
-				SessionData: sessiondatapb.SessionData{
-					VectorizeMode: vectorizeMode,
-				},
-				LocalOnlySessionData: sessiondatapb.LocalOnlySessionData{
-					DistSQLMode: sessiondatapb.DistSQLOn,
-				},
+		ie := execCfg.InternalExecutorFactory(ctx, &sessiondata.SessionData{
+			SessionData: sessiondatapb.SessionData{
+				VectorizeMode: vectorizeMode,
 			},
-		)
+			LocalOnlySessionData: sessiondatapb.LocalOnlySessionData{
+				DistSQLMode: sessiondatapb.DistSQLOn,
+			},
+		})
 		_, err := ie.ExecEx(
 			execCtx,
 			t.Name(),

